blog/content/post/2021-10-28-secure-jtag-on-imx-soc.md

```diff
@@ -0,0 +1,270 @@
+
+- **JTAG compatible with OpenOCD**: On IMX sites we can read that Secure JTAG is
+supported only by the Lauterbach environment and ARM-DS5 IDE with DSTREAM
+debugger. In 3mdeb we are always trying to use open-source software.
+Unfortunately, OpenOCD doesn't officially support Secure JTAG in i.MX SoC's,
+but I found not accepted yet patch from 2014 which add support for work with SJC
+(System JTAG controller) in a secure mode. This patch work on OpenOCD from
+[official repository](git://git.code.sf.net/p/openocd/code) if you checkout this
+with `970a12aef` hash. Is a chance to apply this in the actual version of
+OpenOCD, but it needs some code changes. As a hardware, i used `ARM-USB-OCD-H`
+from Olimex,
+
+- **System image**: Any Linux distribution with kernel module `nvmem-imx-ocotp`
+- for modify OTP memory.
+
+## How it works
+
+i.MX6 and i.MX8M offers JTAG in three modes:
+
+- **No debug** - security on the highest level, but after setting that
+development and debugging by JTAG port will be unavailable permanently on this
+device,
+
+- **Secure JTAG** - mode described in this article, the best compromise between
+security and development possibilities - only person with access key can use
+JTAG,
+
+- **Enabled JTAG** - default mode, enabled for anyone who has physical access to
+JTAG pins or even only device (is a chance to connect to the JTAG pins if the
+connector isn't soldered).
+
+Secure JTAG mode based on challenge/response mechanism. SoC has saved during
+manufacturing a unique challenge key. User can generate their response 56-bit
+key and program it in JTAG fuse `SJC_RESP`. While trying to access the JTAG
+port, SJC gives a challenge key. We can simplify this to the JTAG controller ID.
+Now user should pass the response key. It is compared with the response stored
+in SoC fuse bits. If keys are the same, JTAG is enabled. Bellow, you can see
+schemat described this mechanism:
+
+![Secure JTAG - how it works](/img/secure-jtag.png)
+
+JTAG mode can be set in `JTAG_SMODE` fuse by values:
+
+| Mode        | JTAG_SMODE |
+|-------------|------------|
+| JTAG enable | 0x0        |
+| JTAG secure | 0x1        |
+| No debug    | 0x2        |
+
+These fuses still can be overwritten until we do not block this possibility -
+the response key can be locked by writing `0x1` to `SJC_RESP_LOCK`, and JTAG
+mode with all other fuses from `BOOT` group by writing `0x3` to `BOOT_CFG_LOCK`
+now they should be protected from writing and overwriting. There is an option to
+set this to locking only override - look at the reference manual for more
+information.
+
+## Setting up
+
+### Using cruible
+
+Run `cruible` on your target and check that you have access to fuse bits from
+userspace:
+```
+user@machine:~# ./crucible -m IMX6UL -r 1 -b 16 read JTAG_SMODE 
+soc:IMX6UL ref:1 otp:JTAG_SMODE op:read addr:0x18 off:22 len:2 val:0x0
+```
+
+If you try to burn fuses on i.MX8MM there can be problems with availabe of these
```

Comment:
When you will be changing that to i.MX8MM lets add link to the issue you post.

Reply:
added